mono/metadata/class.c

```diff
@@ -8089,6 +8089,11 @@ mono_class_is_variant_compatible (MonoClass *klass, MonoClass *oklass, gboolean
 		MonoClass *param1_class = mono_class_from_mono_type (klass_argv [j]);
 		MonoClass *param2_class = mono_class_from_mono_type (oklass_argv [j]);
 
+		if (param1_class->valuetype && param2_class->valuetype) {
```

Comment:
Why do we need this?

Reply:
@alexischr I think this is for `mono/tests/bug-400716.cs` in the `TestEnumerator<byte[], IList<sbyte>>` case, right?  Shouldn't that end up in the array special interfaces branch?

Reply:
Is `sbyte[]` castable to `IList<sbyte>` and `IList<byte>`?

Reply:
@lambdageek @kumpera This is specifically for IEnumerator, which is both a special interface and not implemented by arrays. IEnumerator<sbyte> needs to be assignable from IEnumerator<byte>. Here is an example extracted from `bug-400716.cs`:
```
			object obj = new byte[10];
			IEnumerable<sbyte> dd = (IEnumerable<sbyte>)obj;
			var tx = dd.GetEnumerator ();
			IEnumerator<sbyte> x = (IEnumerator<sbyte>)tx; // <--
			x.MoveNext ();
			sbyte t = x.Current;
```

Reply:
I think `byte[]` needs to implement both `IEnumerable<byte>` and `IEnumerable<sbyte>`, rather than some change to how casting works.

Reply:
On .net if I add `Console.WriteLine (tx.GetType ())`, I see ``SZArrayHelper+SZGenericArrayEnumerator`1[System.Byte]`` and ``SZArrayHelper+SZGenericArrayEnumerator`1[System.SByte]`` from the same `byte[]`.  So it really looks like it has two different `GetEnumerator` methods.

Reply:
@lambdageek  The problem is that mono returns the same underlying type for all Enumerators while dotnet doesn't. This is a tech debt we need to address on our end eventually.